### PR TITLE
check and fix pr2171

### PR DESCRIPTION
**Pull request by qiankunli**

### CODE REVIEW

File: .github/workflows/e2e_spark.yaml [hunk 1]
```diff
@@ -57,6 +59,7 @@ jobs:
         eval $(minikube docker-env)
         minikube kubectl -- create clusterrolebinding serviceaccounts-cluster-admin --clusterrole=cluster-admin --group=system:serviceaccounts || true
         build/sbt -Pvolcano -Pkubernetes -Pkubernetes-integration-tests -Dtest.include.tags=volcano "kubernetes-integration-tests/test"
+        kubectl logs $(kubectl get po -nvolcano-system | grep volcano-scheduler |cut -d" " -f1) -nvolcano-system
       working-directory: ${{ github.workspace }}/spark
```
Comment (Yikun) [inline .github/workflows/e2e_spark.yaml:63]:
```suggestion
      working-directory: ${{ github.workspace }}/spark
    - name: Collect Volcano logs
      if: failure()
      run: |
        kubectl logs $(kubectl get po -nvolcano-system | grep volcano-scheduler |cut -d" " -f1) -nvolcano-system
        kubectl get pod -A
```

File: .github/workflows/e2e_spark.yaml [hunk 2]
```diff
@@ -57,7 +59,13 @@ jobs:
...
+    - name: Collect Volcano logs
+        if: failure()
```
Comment (Yikun) [inline .github/workflows/e2e_spark.yaml:65]: intention

**Comment by Yikun** [comment]
```

I0505 01:35:24.585274       1 allocate.go:43] Enter Allocate ...
I0505 01:35:24.585307       1 allocate.go:96] Try to allocate resource to 1 Namespaces
I0505 01:35:24.585318       1 allocate.go:163] Try to allocate resource to Jobs in Namespace <default> Queue <queue-2u-3g>
I0505 01:35:24.585332       1 allocate.go:197] Try to allocate resource to 1 tasks of Job <default/spark-8618f8bd8c8e43d9b1af8e778a16ff56-podgroup>
I0505 01:35:24.585343       1 allocate.go:210] There are <1> nodes for Job <default/spark-8618f8bd8c8e43d9b1af8e778a16ff56-podgroup>
I0505 01:35:24.585466       1 allocate.go:242] Binding Task <default/spark-test-app-813b5aaf02f6492d863ef6a6576062fd-driver-1> to node <docker-desktop>
I0505 01:35:24.585531       1 statement.go:268] After allocated Task <default/spark-test-app-813b5aaf02f6492d863ef6a6576062fd-driver-1> to Node <docker-desktop>: idle <cpu 4150.00, memory 7530921984.00, hugepages-1Gi 0.00, hugepages-2Mi 0.00, hugepages-32Mi 0.00, hugepages-64Ki 0.00>, used <cpu 1850.00, memory 1728053248.00>, releasing <cpu 0.00, memory 0.00>
I0505 01:35:24.585737       1 statement.go:286] Allocating operations ...
I0505 01:35:24.585762       1 statement.go:380] Committing operations ...
I0505 01:35:24.585810       1 allocate.go:159] Namespace <default> have no queue, skip it
I0505 01:35:24.585821       1 allocate.go:284] Leaving Allocate ...
I0505 01:35:24.585825       1 backfill.go:40] Enter Backfill ...
I0505 01:35:24.585828       1 backfill.go:90] Leaving Backfill ...
I0505 01:35:24.618456       1 cache.go:723] bind ok, latency 22.432417ms
I0505 01:35:24.638978       1 cache.go:249] Updating pod condition for default/spark-test-app-813b5aaf02f6492d863ef6a6576062fd-driver-3 to (PodScheduled==False)
I0505 01:35:24.653351       1 cache.go:249] Updating pod condition for default/spark-test-app-813b5aaf02f6492d863ef6a6576062fd-driver-2 to (PodScheduled==False)
I0505 01:35:24.696574       1 session.go:196] Close Session 8de09815-efbe-41fa-90d7-4eabfede321a
I0505 01:35:25.697236       1 cache.go:1069] There are <3> Jobs, <2> Queues and <1> Nodes in total for scheduling.
I0505 01:35:25.697256       1 session.go:174] Open Session 7d53147e-d79f-4690-a0de-948e1036be64 with <3> Job and <2> Queues
E0505 01:35:25.697697       1 runtime.go:76] Observed a panic: resource is not sufficient to do operation: <cpu 1000.00, memory 1476395008.00> sub <cpu 2000.00, memory 0.00>
goroutine 117 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x14cde20, 0x40006d83e0})
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa0
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0})
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x90
panic({0x14cde20, 0x40006d83e0})
	/opt/homebrew/Cellar/go/1.17.8/libexec/src/runtime/panic.go:1038 +0x224
volcano.sh/volcano/pkg/scheduler/util/assert.Assert(0x0, {0x40005ac4d0, 0x6e})
	/Users/yikun/code/volcano/pkg/scheduler/util/assert/assert.go:33 +0x13c
volcano.sh/volcano/pkg/scheduler/util/assert.Assertf(0x0, {0x1888324, 0x39}, {0x40008074a8, 0x2, 0x2})
	/Users/yikun/code/volcano/pkg/scheduler/util/assert/assert.go:43 +0x78
volcano.sh/volcano/pkg/scheduler/api.(*Resource).Sub(0x40007940a0, 0x40007940c0)
	/Users/yikun/code/volcano/pkg/scheduler/api/resource_info.go:234 +0x108
volcano.sh/volcano/pkg/scheduler/api.(*JobInfo).GetElasticResources(0x40007165a0)
	/Users/yikun/code/volcano/pkg/scheduler/api/job_info.go:486 +0x2e4
volcano.sh/volcano/pkg/scheduler/plugins/proportion.(*proportionPlugin).OnSessionOpen(0x4000acde40, 0x40005e0540)
	/Users/yikun/code/volcano/pkg/scheduler/plugins/proportion/proportion.go:158 +0x1f4
volcano.sh/volcano/pkg/scheduler/framework.OpenSession({0x1afadd8, 0x4000466b40}, {0x400065e870, 0x2, 0x2}, {0x0, 0x0, 0x0})
	/Users/yikun/code/volcano/pkg/scheduler/framework/framework.go:43 +0x244
volcano.sh/volcano/pkg/scheduler.(*Scheduler).runOnce(0x40006b9c20)
	/Users/yikun/code/volcano/pkg/scheduler/scheduler.go:105 +0x240
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x400067a4b0)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x64
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x400067a4b0, {0x1a7d860, 0x400065e990}, 0x1, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x94
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x400067a4b0, 0x3b9aca00, 0x0, 0x1, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x88
k8s.io/apimachinery/pkg/util/wait.Until(0x400067a4b0, 0x3b9aca00, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x48
created by volcano.sh/volcano/pkg/scheduler.(*Scheduler).Run
	/Users/yikun/code/volcano/pkg/scheduler/scheduler.go:91 +0x17c
panic: resource is not sufficient to do operation: <cpu 1000.00, memory 1476395008.00> sub <cpu 2000.00, memory 0.00> [recovered]
	panic: resource is not sufficient to do operation: <cpu 1000.00, memory 1476395008.00> sub <cpu 2000.00, memory 0.00>

goroutine 117 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0})
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x10c
panic({0x14cde20, 0x40006d83e0})
	/opt/homebrew/Cellar/go/1.17.8/libexec/src/runtime/panic.go:1038 +0x224
volcano.sh/volcano/pkg/scheduler/util/assert.Assert(0x0, {0x40005ac4d0, 0x6e})
	/Users/yikun/code/volcano/pkg/scheduler/util/assert/assert.go:33 +0x13c
volcano.sh/volcano/pkg/scheduler/util/assert.Assertf(0x0, {0x1888324, 0x39}, {0x40008074a8, 0x2, 0x2})
	/Users/yikun/code/volcano/pkg/scheduler/util/assert/assert.go:43 +0x78
volcano.sh/volcano/pkg/scheduler/api.(*Resource).Sub(0x40007940a0, 0x40007940c0)
	/Users/yikun/code/volcano/pkg/scheduler/api/resource_info.go:234 +0x108
volcano.sh/volcano/pkg/scheduler/api.(*JobInfo).GetElasticResources(0x40007165a0)
	/Users/yikun/code/volcano/pkg/scheduler/api/job_info.go:486 +0x2e4
volcano.sh/volcano/pkg/scheduler/plugins/proportion.(*proportionPlugin).OnSessionOpen(0x4000acde40, 0x40005e0540)
	/Users/yikun/code/volcano/pkg/scheduler/plugins/proportion/proportion.go:158 +0x1f4
volcano.sh/volcano/pkg/scheduler/framework.OpenSession({0x1afadd8, 0x4000466b40}, {0x400065e870, 0x2, 0x2}, {0x0, 0x0, 0x0})
	/Users/yikun/code/volcano/pkg/scheduler/framework/framework.go:43 +0x244
volcano.sh/volcano/pkg/scheduler.(*Scheduler).runOnce(0x40006b9c20)
	/Users/yikun/code/volcano/pkg/scheduler/scheduler.go:105 +0x240
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x400067a4b0)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x64
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x400067a4b0, {0x1a7d860, 0x400065e990}, 0x1, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x94
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x400067a4b0, 0x3b9aca00, 0x0, 0x1, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x88
k8s.io/apimachinery/pkg/util/wait.Until(0x400067a4b0, 0x3b9aca00, 0x4000114a80)
	/Users/yikun/code/volcano/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x48
created by volcano.sh/volcano/pkg/scheduler.(*Scheduler).Run
	/Users/yikun/code/volcano/pkg/scheduler/scheduler.go:91 +0x17c
```

**Comment by k82cn** [comment]
/hold

IMO, it's better to check result or status of pod instead of scheduler log.

**Comment by Yikun** [comment]
```
➜  ~ k get pod
NAME                                                       READY   STATUS    RESTARTS   AGE
spark-pi-7fe5078091f14eed-exec-1                           0/1     Pending   0          52m
spark-test-app-d7678016659446a8b5254d728b895689-driver-1   1/1     Running   0          52m
spark-test-app-d7678016659446a8b5254d728b895689-driver-2   0/1     Pending   0          52m
spark-test-app-d7678016659446a8b5254d728b895689-driver-3   0/1     Pending   0          52m
```

the expected pod status is job completed one by one rather than stucking

**Comment by qiankunli** [comment]
@Yikun @k82cn 

```
// volcano/pkg/scheduler/api/job_info.go
func (ji *JobInfo) GetElasticResources() *Resource {
	if ji.Allocated.LessEqual(ji.GetMinResources(), Zero) {
		return EmptyResource()
	}
	return ji.Allocated.Clone().Sub(ji.GetMinResources())
}
// volcano/pkg/scheduler/api/resource_info.go
func (r *Resource) Sub(rr *Resource) *Resource {
	assert.Assertf(rr.LessEqual(r, Zero), "resource is not sufficient to do operation: <%v> sub <%v>", r, rr)
	return r.sub(rr)
}
```

error `resource is not sufficient to do operation: <cpu 1000.00, memory 1476395008.00> sub <cpu 2000.00, memory 0.00>

we should make sure `job.Allocated ` LessEqual `job.GetMinResources`, normally ` job.GetMinResources = pod resources * minMember`, but in the spark case, the PodGroup specific the minResources  explicitly,  so that ` job.Allocated.cpu <  job.GetMinResources.cpu ` but ` job.Allocated.mem > job.GetMinResources.mem`

we can change LessEqual to  LessEqualPartly to fit the problem.

```
// volcano/pkg/scheduler/api/job_info.go
func (ji *JobInfo) GetElasticResources() *Resource {
	if ji.Allocated.LessEqualPartly(ji.GetMinResources(), Zero) {
		return EmptyResource()
	}
	return ji.Allocated.Clone().Sub(ji.GetMinResources())
}
```

**Comment by Yikun** [comment]
BTW, below are complete yaml test to help us complete test.

<details><summary>min-cpu-driver.yaml</summary>

```yaml
apiVersion: scheduling.volcano.sh/v1beta1
kind: Queue
metadata:
  name: queue-2u
spec:
  weight: 1
  capability:
    cpu: "2"
---
apiVersion: scheduling.volcano.sh/v1beta1
kind: PodGroup
metadata:
  namespace: default
  name: test-pg1
spec:
  queue: queue-2u
  minMember: 1
  minResources:
    cpu: "2"
---
apiVersion: scheduling.volcano.sh/v1beta1
kind: PodGroup
metadata:
  namespace: default
  name: test-pg2
spec:
  queue: queue-2u
  minMember: 1
  minResources:
    cpu: "2"
---
apiVersion: scheduling.volcano.sh/v1beta1
kind: PodGroup
metadata:
  namespace: default
  name: test-pg3
spec:
  queue: queue-2u
  minMember: 1
  minResources:
    cpu: "2"
---
apiVersion: v1
kind: Pod
metadata:
  name: driver-1
  namespace: default
  annotations:
    scheduling.k8s.io/group-name: test-pg1
spec:
  schedulerName: volcano
  containers:
  - name: memory-demo-ctr
    image: ubuntu
    resources:
      requests:
        cpu: "1"
        memory: "1408Mi"
      limits:
        memory: "1408Mi"
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo hello; sleep 10;done"]
---
apiVersion: v1
kind: Pod
metadata:
  name: driver-2
  namespace: default
  annotations:
    scheduling.k8s.io/group-name: test-pg2
spec:
  schedulerName: volcano
  containers:
  - name: memory-demo-ctr
    image: ubuntu
    resources:
      requests:
        cpu: "1"
        memory: "1408Mi"
      limits:
        memory: "1408Mi"
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo hello; sleep 10;done"]
---
apiVersion: v1
kind: Pod
metadata:
  name: driver-3
  namespace: default
  annotations:
    scheduling.k8s.io/group-name: test-pg3
spec:
  schedulerName: volcano
  containers:
  - name: memory-demo-ctr
    image: ubuntu
    resources:
      requests:
        cpu: "1"
        memory: "1408Mi"
      limits:
        memory: "1408Mi"
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo hello; sleep 10;done"]
```


</details>

<details><summary>min-cpu-exec.yaml</summary>

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: exec-1
  namespace: default
  annotations:
    scheduling.k8s.io/group-name: test-pg1
spec:
  schedulerName: volcano
  containers:
  - name: memory-demo-ctr
    image: ubuntu
    resources:
      requests:
        cpu: "1"
        memory: "1408Mi"
      limits:
        memory: "1408Mi"
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo hello; sleep 10;done"]
```

</details>

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with volcano scheduler (12 seconds, 286 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minCPU (35 seconds, 404 milliseconds)
[info] - SPARK-38187: Run SparkPi Jobs with minMemory (32 seconds, 268 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (16 seconds, 124 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (22 seconds, 173 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (17 seconds, 162 milliseconds)
```

Spark IT recovered by this PR, LGTM!

**Review by Yikun — APPROVED** [review]

**Pull request merged**